app/src/main/kotlin/com/mapbox/vision/teaser/MainActivity.kt

```diff
@@ -179,4 +639,44 @@ class MainActivity : BaseTeaserActivity(), ReplayModeFragment.OnSelectModeItemLi
             ArReplayNavigationActivity.start(this, sessionPath)
         }
     }
+
+    override fun onRequestPermissionsResult(
+            requestCode: Int,
+            permissions: Array<out String>,
+            grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (allPermissionsGranted() && requestCode == PERMISSIONS_REQUEST_CODE) {
+            onPermissionsGranted()
+        }
+    }
+
+    private fun allPermissionsGranted(): Boolean {
+        for (permission in getRequiredPermissions()) {
+            if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
+                // PERMISSION_FOREGROUND_SERVICE was added for targetSdkVersion >= 28, it is normal and always granted, but should be added to the Manifest file
+                // on devices with Android < P(9) checkSelfPermission(PERMISSION_FOREGROUND_SERVICE) can return PERMISSION_DENIED, but in fact it is GRANTED, so skip it
+                // https://developer.android.com/guide/components/services#Foreground
+                if (permission == PERMISSION_FOREGROUND_SERVICE) {
+                    continue
+                }
+                return false
+            }
+        }
+        return true
+    }
+
+    private fun getRequiredPermissions(): Array<String> {
```

Comment:
Suggestion : remove all code for manual permission management and use Android Core helpers, https://docs.mapbox.com/android/core/overview/. 
Up to you whether to use it in this (or following PRs), I've just thought about it long ago and seems this refactor is a good time to apply it

Reply:
I'd prefer to do it later after all current modifications, 
Added issue with notes
https://github.com/mapbox/mapbox-vision-android-teaser/issues/191